src/core/tsi/s2a/record_protocol/s2a_crypter.cc

```diff
@@ -598,6 +602,231 @@ grpc_status_code s2a_encrypt(s2a_crypter* crypter, uint8_t* plaintext,
+  GPR_ASSERT(crypter != nullptr);
+  GPR_ASSERT(plaintext_bytes_written != nullptr);
+  size_t max_plaintext_size;
+  grpc_status_code plaintext_status =
+      s2a_max_plaintext_size(crypter, protected_vec, protected_vec_size,
+                             &max_plaintext_size, error_details);
+  if (plaintext_status != GRPC_STATUS_OK) {
+    return INTERNAL_ERROR;
+  }
+  GPR_ASSERT(static_cast<int>(unprotected_vec.iov_len) >= max_plaintext_size);
+  uint8_t* header = reinterpret_cast<uint8_t*>(record_header.iov_base);
+  const uint16_t wire_version = static_cast<uint16_t>(TLS1_2_VERSION);
+  size_t payload_size = get_total_length(protected_vec, protected_vec_size);
+  size_t tag_size;
+  grpc_status_code tag_status = s2a_tag_size(crypter, &tag_size, error_details);
+  if (tag_status != GRPC_STATUS_OK) {
+    return INTERNAL_ERROR;
+  }
+  if (payload_size >
+      SSL3_RT_MAX_PLAIN_LENGTH + tag_size + /** record type **/ 1) {
+    /** If the plaintext size exceeds the max allowed, the TLS 1.3 RFC demands
+     *  that the record protocol returns a "record overflow" alert and that the
+     *  connection be terminated; for more details, see
+     *  https://tools.ietf.org/html/rfc8446#section-5.2 . **/
+    *error_details = gpr_strdup(kS2ARecordExceedMaxSize);
+    return ALERT_RECORD_OVERFLOW;
+  }
+  size_t expected_payload_size = (header[3] << 8) + header[4];
+  if (record_header.iov_len != SSL3_RT_HEADER_LENGTH ||
+      header[0] != SSL3_RT_APPLICATION_DATA ||
+      header[1] != (wire_version >> 8) || header[2] != (wire_version & 0xff) ||
+      payload_size != expected_payload_size) {
+    *error_details = gpr_strdup(kS2AHeaderIncorrectFormat);
+    return INVALID_RECORD;
+  }
+
+  s2a_decrypt_status decrypt_status = s2a_decrypt_payload(
+      crypter, record_header, protected_vec, protected_vec_size,
+      unprotected_vec, plaintext_bytes_written, error_details);
+  if (decrypt_status != OK) {
+    return decrypt_status;
+  }
+
+  grpc_status_code increment_status =
+      increment_sequence(crypter->in_connection);
+  if (increment_status != GRPC_STATUS_OK) {
+    return INTERNAL_ERROR;
+  }
+
+  uint8_t* plaintext = reinterpret_cast<uint8_t*>(unprotected_vec.iov_base);
+  /** At this point, the |s2a_decrypt_payload| method has written
+   *  |*plaintext_bytes_written| bytes to |plaintext|, and these bytes are of the form
+   *    (plaintext) + (record type byte) + (trailing zeros).
+   *  These trailing zeros should be ignored, so we will search from one end of
+   *  the |plaintext| buffer until we find the first nonzero trailing byte,
+   *  which must be the record type.
+   *
+   *  Note that this TLS 1.3 implementation does not add padding by zeros when
+   *  constructing a TLS 1.3 record; nonetheless, |s2a_decrypt_payload| must be
+   *  able to parse a TLS 1.3 record that does have padding by zeros. **/
+  size_t i;
+  for (i = *plaintext_bytes_written - 1; i >= 0; i--) {
```

Comment:
Does this overflow if you don't find a non-zero byte since i is unsigned?

Reply:
Good find. If you don't find a nonzero byte, it will overflow. I have changed the check on `i` to reflect this.